Upgrade generated parse_line() to accept constant SBuf

rousskov opened the pull request:
Also applied recent commit 314e430 "only SP and HT" logic when parsing
filenames in preprocessor `include` directives and operands in
preprocessor `if` statements. Prior to this change, in these contexts,
Squid also accepted ASCII VT and FF characters as well as any other
locale-specific characters classified as space by isspace(3).

Other changes attempt to preserve existing squid.conf grammar
idiosyncrasies. A lot of squid.conf parsing code needs to be upgraded;
these changes focus on parse_line() because upcoming reconfiguration
improvements require that particular function to accept SBuf.

Also fixed ConfigParser SetCfgLine() method and CfgLine data member
descriptions. These members are not about the whole squid.conf line, and
not every squid.conf line goes through these members: These members are
dedicated to tokens located after a directive name on a directive line.
A lot of related code needs to be polished, but we fixed these docs
because this change affects all SetCfgLine() callers and CfgLine users.

rousskov committed:
Upgrade generated parse_line() to accept SBuf

These changes attempt to preserve existing squid.conf grammar
idiosyncrasies. A lot of squid.conf parsing code needs to be upgraded;
these changes focus on parse_line() because upcoming reconfiguration
improvements require that particular function to accept SBuf.

Also fixed ConfigParser SetCfgLine() method and CfgLine data member
descriptions. These members are not about the whole squid.conf line, and
not every squid.conf line goes through these members: These members are
dedicated to tokens located after a directive name on a directive line.
A lot of related code needs to be polished, but we fixed these docs
because this change affects all SetCfgLine() callers and CfgLine users.

rousskov commented on src/base/CharacterSet.cc:
```diff
@@ -168,3 +168,18 @@ CharacterSet::RFC3986_UNRESERVED()
+    const auto creator = []() -> auto {
+        const auto chars = new CharacterSet("libcSpace");
+        for (size_t idx = 0; idx < 256; ++idx) {
+            if (xisspace(idx))
```
I added this CharacterSet to preserve locale-specific behavior of existing squid.conf parsing code (that used to call xisspace() via trim_trailing_ws(), FindStatement(), and when parsing `include` preprocessor directive). We may decide to stop being locale-sensitive in these parsing contexts (e.g., like we stopped relying on isspace() when validating Content-Length value in commit 9fcf5206), but let's not make that decision in _this_ behavior-preserving PR.

yadij replied:
Lets do make that decision before baking support for indeterminate invisible characters into squid.conf.
It is an easy one to make, 

FTR, the Squid community is **already** sharing `squid.conf` snippets widely with the assumption that the contents **do not** contain any locale-specific characters.

So ... please drop this logic. Only support the portable `CharacterSet::WSP` set for whitespace characters.

rousskov replied:
> Lets do make that decision before baking support for indeterminate invisible characters into squid.conf.

The implied assumption that this PR is "baking support for indeterminate invisible characters into squid.conf" is false: This PR preserves official code behavior in this area. In other words, that (questionable) behavior is already as "baked in" as it will be after this PR. I strongly recommend changing this in a followup PR. Do you insist on _this_ PR changing Squid behavior in this area?

yadij replied:
Adding complicated code, knowing it will be erased in a (nearly immediate) followup? that is a bad practice @rousskov.

Just omitting the new complex code solves several issues and makes this PR a fair amount cleaner.

rousskov replied:
> Adding complicated code, knowing it will be erased in a (nearly immediate) followup? that is a bad practice

This PR does not add complicated code. A bit of added code will probably be eventually erased, but that is not a big deal _and_ a followup PR (if done right) will have to modify a lot more code, code that this PR does not and should not touch.

> Just omitting the new complex code solves several issues and makes this PR a fair amount cleaner.

Not really. On the other hand, insisting on those out of scope changes (with significant side effects!) does delay this innocent PR even more. It is sad that straightforward, correct, useful, small, and functionality-preserving Squid changes are so costly!

rousskov replied:
> please drop this logic. Only support the portable `CharacterSet::WSP` set for whitespace characters.

Done in #1858 to keep serious behavior changes (and associated noise) outside this "upgrade to SBuf" PR.

rousskov replied:
> > please drop this logic. Only support the portable `CharacterSet::WSP` set for whitespace characters.
> 
> Done in #1858 to keep serious behavior changes (and associated noise) outside this "upgrade to SBuf" PR.

And those official changes are now merged into this PR branch. I have updated PR description accordingly.

When resolving conflicts, I found two more relevant space handling cases. One of them (former strwordtok() in `include` filename parsing) could have been handled in commit 314e430, but I missed it because it did not call isspace() directly. Another (parsing integers in `if` conditions) could not be handled earlier because strtoll() and friends are using isspace() implicitly (and we could not remove those calls without upgrading to Tokenizer which is done in this PR).

@yadij, I hope this master merge addresses your change request. Please review.

rousskov commented on src/parser/Tokenizer.cc:
```diff
@@ -333,3 +333,60 @@ Parser::Tokenizer::udec64(const char *description, const SBuf::size_type limit)
+}
+
+std::optional<SBuf>
+NextWordWhileRemovingDoubleQuotesAndBackslashesInsideThem(Parser::Tokenizer &tk)
```
I am using this very long name because I assume that the underlying syntax here does not have a well-known name (it is effectively defined by existing strwordtok() _implementation_ that this code is based on) and to automatically flag all callers for extra scrutiny: If the caller should use a well-known syntax instead, they should not call this function.

Just like with libcSpace(): We may (and probably should!) change callers grammar to some well-known alternative, but let's not do that in this behavior-preserving PR.

If my assumption about the lack of a well-known name for this syntax is wrong, please let me know what the correct name is, and I will rename this function in his PR, of course.

yadij replied:
Er, this is called "un-escaping a quoted string"

rousskov replied:
> Er, this is called "un-escaping a quoted string"

Yes, it is. A lot of things can be (and are!) called that. This function name should be more precise IMO, for the reasons detailed earlier.

yadij replied:
Alex: I am using this very long name because I assume that the underlying syntax here does not have a well-known name

The well-known name is "un-escaping a quoted-string". Therefore your fundamental assumption was wrong. Reasons based on a wrong assumption are not valid.

Despite currently lacking support for some characters, the escaping in `squid.conf` is expected to follow ANSI-C rules. To match what administrators should already be familiar with from the Bourne-shell (or equivalent) tools. see https://www.gnu.org/software/bash/manual/bash.html#ANSI_002dC-Quoting

rousskov replied:
> The well-known name is "un-escaping a quoted-string".

Yes, it is, but that well-known name is too generic -- a lot of things can be (and are!) called that. This function name should be more precise IMO, for the reasons detailed earlier.

> Despite currently lacking support for some characters, the escaping in squid.conf is expected to follow ANSI-C rules. To match what administrators should already be familiar with from the Bourne-shell (or equivalent) tools. see https://www.gnu.org/software/bash/manual/bash.html#ANSI_002dC-Quoting

This function does not match ANSI C escaping rules. The lack of support for some escape sequences alone would be enough to reach that important conclusion (and justify a special name), but the existing differences do not stop there and are quite fundamental (rather than superficial as in "we just have not implemented that particular escape sequence yet but ban it in input"). Here are a few of those other differences:

* Unsupported escape sequences are treated differently than required by ANSI C and than implemented for quoted strings in bash: In most cases, Squid _removes_ backslash and leaves the escaped character. ANSI C requires diagnostics for unsupported escape characters. Bash _leaves_ backslash untouched in the quoted string if the escaped character is unsupported. The difference between Squid's `foo.bar` and bash's `foo\.bar` result can be important.
* Squid treats escape sequences specially only _inside_ double quoted strings. Bash processes escape sequences in unquoted/bare strings as well. Again, the difference can be important.

However, if you insist on renaming this function, then please suggest an exact name that you would like to see here, as explicitly requested more than a year ago on this comment thread. We may be able to make progress here without discussing these escaping differences.

kinkie replied:
How about neither-nor? TBH This discussion feels overblown. This function has exactly one customer, in cache_cf.

I propose that make that explicit, call it ConfigNextQuotedWord, document the behaviour and use case in a doxygen comment and move on

rousskov replied:
> This function has exactly one customer, in cache_cf.

True, but this function is currently meant (and explicitly documented!) to help get rid of all legacy strwordtok() calls as we upgrade code to use SBuf and friends.

If you think it will help to merge this PR, please let me know and I will make this function specific to its current two callers in `cache_cf.cc` instead of trying to provide a replacement that other/future callers may use. In this case, the new function can be placed inside cache_cf.cc and called something like NextIncludeFilenameGlob() or even NextIncludeFilename(). Doing so will simplify things a bit (for this PR) while making complicating future work. Should I do that?

> I propose [to] call it ConfigNextQuotedWord

If you insist, I will use that name, but I am concerned that the callers do not require a quoted word and may unescape many quoted strings at once. After reading your naming suggestion, I made the first fact more explicit in a new PR branch commit 1203a68a. The other fact is more-or-less clear from the new function name.

For now, I renamed this function to NextWordWhileUnescapingQuotedStrings() in hope that a shorter name based on Amos' `this is called "un-escaping a quoted string"` [assertion](https://github.com/squid-cache/squid/pull/1840#discussion_r1659405596) allows us to make progress. That renaming was done in PR branch commit f0f2db68.

I also copied special`\n` and `\r` and trailing space handling from strwordtok() to better match the documented strwordtok() replacement goal and improved descriptions. I also merged master/v8 changes to get CI fixes in case this PR is finally unblocked. Here is the list of primary recent changes:

* 55b98a92: Merge master/v8 to get recent CI fixes
* 1203a68a: fixup: Explicitly document that the next token may not start with a quote
* 777e2df4: fixup: Documented differences in space treatment
* fc56ad7c: fixup: Match more aspects of strwordtok() behavior
* f0f2db68: fixup: Renamed NextWordWhileRemovingDoubleQuotesAndBackslashesInsideThem()

> This discussion feels overblown

I agree, and I hope that shared feeling helps us make progress.

rousskov reviewed (commented):
This review annotates PR changes but does not request any adjustments.

eduard-bagdasaryan commented on src/cache_cf.cc:
```diff
@@ -522,46 +529,47 @@ parseOneConfigFile(const char *file_name, unsigned int depth)
             continue;
 
-        const char* append = tmp_line_len ? skip_ws(config_input_line) : config_input_line;
+        wholeLine.append(config_input_line);
```
Don't we need skip_ws() here as before?

eduard-bagdasaryan commented on src/cache_cf.cc:
```diff
@@ -298,139 +299,144 @@ skip_ws(const char* s)
-    return (end != str) && (*end == '\0'); // returns true if string contains nothing except number
+    auto numberParser = Parser::Tokenizer(ExtractToken(description, tk));
+    int64_t result = 0;
+    if (!numberParser.int64(result, 0, true))
```
This default '0' (guess) does not work in my configuration `if ${process_number} = 1`:
```
FATAL: bad configuration: cannot parse left-hand operand of an equality condition
    exception location: cache_cf.cc(413) ExtractNumber
```
Parser::Tokenizer() gets zero '0' token but int64() treats it as octal notation.

rousskov replied:
> Parser::Tokenizer() gets zero '0' token but int64() treats it as octal notation.

Indeed. Glad you found this Tokenizer bug! I have posted a fix at #1842. I am marking this PR as blocked on that PR.

rousskov replied:
> Glad you found this Tokenizer bug! I have posted a fix at #1842. I am marking this PR as blocked on that PR.

I have merged that fix and removed S-waiting-for-PR label.

eduard-bagdasaryan commented on src/cache_cf.cc:
```diff
@@ -298,139 +299,144 @@ skip_ws(const char* s)
-    }
-    fatalf("Unable to evaluate expression '%s'\n", expr);
-    return false; // this place cannot be reached
+    const auto op = ExtractToken("equality sign in an equality condition", tk);
```
New parsing code became a bit touchy - e.g., valid old configurations with no spaces `if ${process_number}=1` or two spaces `if ${process_number}  =  1` are not allowed. Are we going to adopt more rigid parsing rules here?

rousskov replied:
> Are we going to adopt more rigid parsing rules here?

No, we should preserve the original grammar (to the extent it is practical). Thank you for flagging this change/bug! I hope I fixed it (as of branch commit 6d897ce0). I also added a few test cases (but they can be improved further by adding SMP macros).

rousskov replied:
> I also added a few test cases (but they can be improved further by adding SMP macros).

I now added an SMP macro test and temporary workarounds for Tokenizer::int64() bugs (so that the tests pass while we wait for PR 1842).

rousskov committed:
fixup: Do not insist on space around pp condition operator

    if 0=1

Our preprocessor grammar is not based on space-separated tokens!

rousskov committed:
fixup: Added a preprocessor test

Condition "00=000" will need to be replaced with "0=000"
when int64() bug is fixed.

rousskov committed:
fixup: Simplified and polished preprocessor parsing code

rousskov committed:
fixup: More temporary workarounds for Tokenizer::int64() bugs

rousskov committed:
Add a weak test of SMP configuration macro expansion

Better than nothing. We could add a stronger test that validates
equality, but it would make this configuration file inapplicable to SMP
Squids. IMO, we should not do that without also developing the
corresponding instruction (and adding that to the instructions file).

rousskov commented on test-suite/squidconf/pp-ifs.conf:
```diff
@@ -0,0 +1,27 @@
+
+if true
+    if 10 = 0xA
+        if 0${process_number} = 077777777
+        else
+            if 00=000
+                if  -00= +00
```
When Tokenizer::int64() can parse "0", most leading zeros will be gone:

```suggestion
        if ${process_number} = 9223372036854775807
        else
            if 0=000
                if  -0= +0
```

rousskov replied:
> When Tokenizer::int64() can parse "0", most leading zeros will be gone

I merged master/v7 commit 8cc5e88c into the branch and removed those distracting zeros as planned (branch commit 2176dd8a).

eduard-bagdasaryan commented on src/base/CharacterSet.cc:
```diff
@@ -168,3 +168,18 @@ CharacterSet::RFC3986_UNRESERVED()
+{
+    const auto creator = []() -> auto {
+        const auto chars = new CharacterSet("libcSpace");
+        for (size_t idx = 0; idx < 256; ++idx) {
```
We could std::numeric_limits<unsigned char>::max() instead of the numeric constant.

rousskov replied:
Done in 4a565ed3. Technically, there were _two_ explicit numeric constants, so I converted both.

eduard-bagdasaryan reviewed (approved):
LGTM

rousskov committed:
fixup: Prefer std::numeric_limits to unnamed constants

Also fix #include order for STL headers.

Context:
https://github.com/squid-cache/squid/pull/1840#pullrequestreview-2123991179

yadij commented on src/ConfigParser.cc:
```diff
@@ -561,6 +562,20 @@ ConfigParser::rejectDuplicateDirective()
+ConfigParser::openDirective(const SBuf &line)
+{
+    Parser::Tokenizer tk(line);
+    SBuf directiveName; // TODO: Upgrade cfg_directive to SBuf and set it here.
```
IMO `cfg_directive` should be removed in favour of a `ConfigParser` member, not just type changed.

rousskov replied:
This change request is a duplicate of https://github.com/squid-cache/squid/pull/1840#discussion_r1649152617

yadij commented on src/ConfigParser.cc:
```diff
@@ -561,6 +562,20 @@ ConfigParser::rejectDuplicateDirective()
+    Parser::Tokenizer tk(line);
+    SBuf directiveName; // TODO: Upgrade cfg_directive to SBuf and set it here.
+    static const auto spaceChars = CharacterSet("w_space", w_space);
+    static const auto directiveChars = spaceChars.complement("squid.conf directive name");
+    const auto found = tk.prefix(directiveName, directiveChars);
```
This introduces a large number of future problems with UTF character encodings and invisible ASCII characters.

Please keep the parser restricted to just the set of characters we currently use in directive names. We can add other specials in future as-needed.
```suggestion
    static const auto directiveChars = CharacterSet("squid.conf directive name","_") + CharacterSet::DIGIT + CharacterSet::ALPHA;
    const auto found = tk.prefix(directiveName, directiveChars);
```

rousskov replied:
> This introduces a large number of future problems with UTF character encodings and invisible ASCII characters.

This code does not _introduce_ those problems: "UTF character encodings and invisible ASCII characters" are handled essentially the same way by current master/v7 code that uses [strtok](https://en.cppreference.com/w/c/string/byte/strtok)() with the same w_space delimiters:

```C++
static int
parse_line(char *buff)
{
        char    *token;
        if ((token = strtok(buff, w_space)) == NULL) 
                return 1;       /* ignore empty lines */
        ConfigParser::SetCfgLine(strtok(nullptr, ""));
    if (!strcmp(token, "broken_vary_encoding"))
        ...
```

This PR _preserves_ that (arguably unwanted) behavior.

> Please keep the parser restricted to just the set of characters we currently use in directive names. We can add other specials in future as-needed.

s/keep/make/ -- the current official parser is not restricted that way.

Are you sure this change should be done in _this_ PR? I will do it if you insist, but it feels like it should be done in a followup PR that is focused on restricting input in these areas. I do support restricting input, but this PR is not about that.

rousskov replied:
> Please keep the parser restricted to just the set of characters we currently use in directive names.

Also, the current directive name parser (in official code and in this PR) is already _effectively_ restricting directive names to "characters we currently use in directive names" -- the parser is written as a series of `if` statements matching recognized directives names and rejecting any other names (including any name with prohibited characters).

The only thing we can add here is a new explicit "[not only you are using an unrecognized directive name, but] you are using a directive name with an invalid character" error. Adding such an error requires more/different code than code suggested in this change request. Such an error will be useful in some rare cases, but adding such an error can clearly be viewed as being outside _this_ "SBuf upgrade" PR scope.

yadij commented on src/cache_cf.cc:
```diff
@@ -298,139 +299,220 @@ skip_ws(const char* s)
+    while (auto path = NextWordWhileRemovingDoubleQuotesAndBackslashesInsideThem(tk)) {
+        if (glob(path->c_str(), globbuf.gl_pathc ? GLOB_APPEND : 0, nullptr, &globbuf) != 0) {
+            const auto xerrno = errno;
+            throw TextException(ToSBuf("Cannot find included configuration file named ", *path, ": ", xstrerr(xerrno)), Here());
```
This is a UI change. Please document in the PR description, or revert to the original FATAL message text.

rousskov replied:
> Please document in the PR description, or revert to the original FATAL message text.

Commit https://github.com/squid-cache/squid/commit/d0e01dd0679a6959155eb0f6045444052a9026a2 restores the original message text.

yadij reviewed (changes requested):
FWIW, I believe you are dong the refactor "wrong", ie. in a way which makes the change steps more difficult and complex than other approaches.

There are a lot of "Also, ..." changes in `src/cache_cf.cc` which are not documented in the PR description.

rousskov committed:
fixup: Undo error text improvement

Context:
https://github.com/squid-cache/squid/pull/1840#discussion_r1649189709

rousskov committed:
fixup: Detail a TODO

Context:
https://github.com/squid-cache/squid/pull/1840#discussion_r1649152617

rousskov commented on src/ConfigParser.cc:
```diff
@@ -561,6 +562,20 @@ ConfigParser::rejectDuplicateDirective()
+ConfigParser::openDirective(const SBuf &line)
+{
+    Parser::Tokenizer tk(line);
+    SBuf directiveName; // TODO: Upgrade cfg_directive to SBuf and set it here.
```
Yes, of course. I adjusted that TODO text in commit https://github.com/squid-cache/squid/commit/896e2f52b6c950ddd0e06c9f2d9c84f98d204837.

rousskov reviewed (commented):
@yadij, I hope I have addressed all change requests. Please revisit the two change requests related to character sets ASAP. I hope those requests were based on the wrong assumption about the current official code, and the corresponding changes should be implemented in a followup PR instead. If you insist, I will address them in this PR, but I do not want this PR wait another ten days for that decision. Thank you.

rousskov committed:
Merged master to resolve a seemingly trivial merge conflict

rousskov committed:
Merge master to resolve conflicts with "only SP/HT" changes

See master/v7 commit 314e430.

rousskov committed:
Switch to "only SP/HT" grammar when parsing include filenames

This change should have been done in master/v7 commit 314e430, but I
missed it because the caller does not use isspace() directly.

This change removes the last use of branch-added
CharacterSet::libcSpace().

rousskov committed:
Removed no-longer-used branch-added CharacterSet::libcSpace()

yadij commented on src/cache_cf.cc:
```diff
@@ -311,139 +312,219 @@ skip_ws(const char* s)
+        if (tk.atEnd())
+            return true;
+
+        if (SkipOptionalSpace(tk)) {
+            RejectTrailingGarbage("preprocessor keyword", keyword, tk);
+            return true;
+        }
+        // e.g., "elseif"
```
Ah, "elseif" is not supported.

However, despite `cf.data.pre` stating that "NOTE: An else-if condition is not supported."  the configuration grammar:
```
   // grammar: "else" space "if" space condition space* END
```
does actually work in current Squid. This new logic will break that. Please fix.

kinkie replied:
I agree. Fixing that behaviour is out of scope here

rousskov committed:
Merge master to get int64() parsing fixes for our test cases

rousskov committed:
fixup: Removed distracting workarounds for int64() zero-parsing bugs

rousskov committed:
fixup: Fixed (or at least improved) test case wording

rousskov reviewed (commented):
@yadij, I believe I have addressed all change requests. Please unblock this PR.

rousskov committed:
Merge branch 'master' into SQUID-984-parse_line-SBuf

yadij commented on src/cache_cf.cc:
```diff
@@ -311,139 +312,219 @@ skip_ws(const char* s)
+/// extracts all leading space characters (if any)
+/// \returns whether at least one character was extracted
+static bool
+SkipOptionalSpace(Parser::Tokenizer &tk)
 {
-    SubstituteMacro(line, len, "${service_name}", service_name.c_str());
-    SubstituteMacro(line, len, "${process_name}", TheKidName.c_str());
-    SubstituteMacro(line, len, "${process_number}", xitoa(KidIdentifier));
+    return tk.skipAll(CharacterSet::WSP);
 }
```
Calling a one-line function is just as much duplicate code as writing its contents in full.

> This function should not be inlined because inlining it will duplicate its code 

"duplicate code" arguments are only relevant to lines which are complex (eg math equations).

One-line functions/methods that simply call another function/method such as this one are called "trampolines" and best practice is to *avoid* use of trampolines whenever possible.
> and obscure its high-level intent/logic in many callers.

The "high level intent" is clear from the syntax and parameters of `skipAll(CharacterSet::WSP);`.
This being a separate function is redundant (as is the new documentation for it).

> 
> In most cases, we should not inline, regardless of the number of lines a function has. There are exceptions to that rule of thumb, of course, but this particular function is not exceptional.

rousskov replied:
> Calling a one-line function is just as much duplicate code as writing its contents in full.

That assertion is false for at least two independent reasons:

* That encapsulated code line contains more details than a function name. It contains the choice of delimiters to use when skipping space.
* Using the same context-specific wrapper function (name) in multiple places carries more information than calling a low level skipAll() method would. Knowing that all these places are related and all of them skip optional space (rather than, say, required space) helps read and maintain code.


> > This function should not be inlined because inlining it will duplicate its code
> 
> "duplicate code" arguments are only relevant to lines which are complex (eg math equations).

That assertion is false because complexity is not the only thing we do not want to duplicate. Duplicating simple decisions is also a negative.

> The "high level intent" is clear from the syntax and parameters of `skipAll(CharacterSet::WSP);`.

Does that call skip optional or required space? What other places should be updated if I want to update WSP in one particular place? Encapsulating that call helps answer these and other questions. Yes, those answers can usually be derived without encapsulation, but at a higher cost (sometimes significantly higher cost!).

> > In most cases, we should not inline, regardless of the number of lines a function has. There are exceptions to that rule of thumb, of course, but this particular function is not exceptional.
> 
> I Disagree, one-line functions are all special-case for that rule of thumb. 

I see no good reason to make one-line functions a special case. The onus is on you to prove such an exception is needed (just stating that it is needed is not enough).

N.B. Absence of an exception to "do not inline unless you have a very good reason" rule of thumb does not mean that one-line functions are _always_ better than duplicated inlined code, of course.

> This one is also an exceptional in addition to that one-line nature:
> 
> FYI: One-line functions/methods that simply call another function/method (such as this one does) are called "trampolines" and best practice is to _avoid_ use of trampolines whenever possible.

The term "[trampoline function](https://en.wikipedia.org/wiki/Trampoline_(computing))" means quite different things in various programming languages/contexts, but it is _not_ used to describe "a function calling another function" in any relevant context I can find. The term is usually related to functions that define/contain (not just call!) other functions or receive other functions as parameters/state. None of that applies here, of course.

yadij commented on src/cache_cf.cc:
```diff
@@ -311,139 +312,219 @@ skip_ws(const char* s)
-        char* lvalue = (char*)xmalloc(equation - expr + 1);
-        xstrncpy(lvalue, expr, equation - expr + 1);
-        trim_trailing_ws(lvalue);
+    // grammar: space* keyword space* END
```
Hmm. right. The wrapper function naming and code structure makes it seem like any skipping done causes a rejection.

The problem this is the opposite. A line **without* spaces after the found token is **not** rejected as garbage. For example;
 * the old parser (as already mentioned) accepted "else if" and treated it as an else-statement with a nested if-statement
 * this logic will accept "elseif" and do the same.
.. both of which are unintentional and broken/undefined behaviour.

rousskov replied:
> A line *_without_ spaces after the found token is **not** rejected as garbage.

That statement is misleading in this context because such a line is **not** accepted either: A line with garbage after letters matching a given keyword K results in IsIfStatementLine(K) returning false. That is correct behavior for that function and the grammar it defines/supports -- preprocessor keyword K was not found. After IsIfStatementLine(K) returns false, other parsing code (possibly including other IsIfStatementLine(X) calls) deals with that line, rejecting it as needed.

> For example; ... this logic will accept "elseif" and [treat it as an else-statement with a nested if-statement]

That example is incorrect or misleading. PR code does not accept "elseif". It correctly rejects "elseif", first as not a preprocessor line and then as an unknown squid.conf directive:

```
2024/11/05 10:51:40| ERROR: unrecognized directive near 'elseif'
2024/11/05 10:51:41| FATAL: configuration failure: Found 1 unrecognized directive(s)
```

If we add a directive named like `elsewhere_services ...` or `iffy_clients_enabled ...`, PR preprocessor code will not need to be modified for Squid to correctly recognize that new directive despite its name starting with the same letters as the preprocessor keyword "else" or "if". Needless to say, the chances of us adding those exact directives are pretty much zero, but I am using them to illustrate that PR code structure is correct and can support future changes such as addition of new directives (or new preprocessor keywords that are more likely to match some directive name prefixes!).

rousskov commented:
@yadij, please unblock this PR. After all change requests were addressed in [July 2024](https://github.com/squid-cache/squid/pull/1840#pullrequestreview-2163818229), this PR has been waiting for your review for more than 5 months, stalling progress and creating a long backlog of changes.

rousskov commented on src/cache_cf.cc:
```diff
@@ -311,139 +312,219 @@ skip_ws(const char* s)
+/// \returns std::nullopt if input does not look like an `include` statement
+/// \returns `include` parameters if input is an `include` statement
+static std::optional<SBuf>
+IsIncludeLine(Parser::Tokenizer tk)
```
> The caller is using it as a parse function, not a look-ahead test.

The caller is using this function as a look-ahead test. That is a fact.

"Using it as a parse function" is _also_ true (for some definition of "using as a parse function"). There is nothing wrong with that kind of dual use, of course -- checking what is ahead is a common parsing step!

> ```c++
>     if (const auto files = IsIncludeLine(tk)) {
>         err_count += parseIncludedConfigFiles(*files, depth + 1);
> ```
> 
> when `IsIncludeLine(tk)` returns the caller expects it to have removed any "include ..." from `tk`

No, the caller does not (and should not) expect that from an IsFoo() function. All those IsFoo() test functions do not modify their input, by design. We have many of them, including:

```
IsIfStatementOpening(Parser::Tokenizer tk)
IsIfStatementLine(const SBuf &keyword, Parser::Tokenizer tk)
IsIncludeLine(Parser::Tokenizer tk)
IsSpace(const char ch)
IsMember(const SBufList & sl, const SBuf &S, const SBufCaseSensitive case_sensitive)
```

> `Tokenizer::skip*()` are already designed not to change the `tk` content if the keyword being searched for is not present.

True, but those skip() functions are designed not to return what they skip because they are meant (and optimized for) callers that do not care about those details. This use case is not a "skip include line, whatever it is" use case.


> IMO, This function should be this:
> 
> ```c++
> static bool
> IsIncludeLine(Parser::Tokenizer tk)
> {
>     const auto keywordInclude = SBuf("include ");
>     if (tk.skip(keywordInclude)) {
>         (void)tk.skipAll(CharacterSet::WSP);
>         return true;
>     }
>     return false;
> }
> ```
> 
> ... and its caller changed to do this:
> 
> ```c++
>     if (IsIncludeLine(tk)) {
>         err_count += parseIncludedConfigFiles(tk, depth + 1);
>         continue; // loop to handle next line.
>     } else {
> ```

PR implementation is better than the above suggestion because:

* PR does not require parseIncludedConfigFiles() to know `include` statement syntax.
* PR does not parse the same `include` statement twice.
* PR separates two very different concerns -- parsing include statements and processing files named by those statements. PR allows for the include statement to evolve without affecting processIncludedFiles(). PR allows processIncludedFiles() to be used outside IsIncludeLine() context.

I am guessing that the suggested version has a bug -- IsIncludeLine() parameter was meant to be a non-constant reference, so that IsIncludeLine() tokenizer modifications last beyond that function call. However, fixing that problem would create new problems (e.g., an isFoo() test should not modify things). There are other problems in the suggested implementation, but I am ignoring them because PR code is going to be better even if they are fixed.

kinkie replied:
TBH I align with Amos on this topic: I do tend to interpret an IsSomething function as a test that does not modify state in any way. However I would not advocate to change the function's contents: it's been developed and tested, that feels unnecessary; I would recommend changing the function's name to reflect its dual use.w It has a limited number of callsites, should not be an excessive burden

rousskov committed:
fixup: Renamed NextWordWhileRemovingDoubleQuotesAndBackslashesInsideThem()

The new name -- NextWordWhileUnescapingQuotedStrings() -- feels too
generic/ordinary for a function with surprising behavior that does not
comply with any escaping standard or convention, but I have failed to
convince others that the original name is a bit better.

rousskov committed:
fixup: Match more aspects of strwordtok() behavior

We claim that the new function is meant to replace legacy strwordtok()
calls. Legacy strwordtok() honors \r and \n escape sequences inside
quoted sections and skips all trailing spaces.

rousskov committed:
fixup: Documented differences in space treatment

... between NextWordWhileUnescapingQuotedStrings() and strwordtok().

These differences stem from 2024 master/v7 commit 314e4304: "Restrict
squid.conf preprocessor space characters to SP and HT".

rousskov committed:
fixup: Explicitly document that the next token may not start with a quote

Context: A ConfigNextQuotedWord() name proposal that could have been
based on the opposite assumption:
https://github.com/squid-cache/squid/pull/1840#discussion_r2188132272

rousskov committed:
Merge master/v8 to get recent CI fixes

rousskov committed:
fixup: Source Format Enforcement after merging master/v8

rousskov reviewed (commented):
> I can see only three open points. Can we close them and move on?

This PR is awaiting Amos review; I cannot just close Amos change requests or dismiss his negative vote. This PR has been in essentially the same state for more than a year now, despite my repeated pleas and reminders.
